feat(avm): Bulk test for bytecode/calldata hashing

MirandaWood opened the pull request:
## Hashing Bulk Test
Originally part of #17224, but I ~separated it out because the test is pretty long (and to help with reviewing!).~

UPDATE: No longer a separate test but included in separate enqueued calls in the main bulk test. This 'only' increases the time by about 10s. The below stats refer to the data hashing tests on their own!

I added ~a new bulk-style test with~ multiple enqueued calls performing calldata and bytecode hashing.

Key points:

- Performs 3x calldata hashes with reads
- Performs 3x protocol contract calls (=> 3x bytecode hashes)
- Currently takes ~20s (10s check_circuit only)

Since we dedupe hashing events, I added different protocol contracts to force the bytecodes to be unique.



### Metrics

```
INFO: avm-data-hashing-test # Public TX Simulation Metrics (ALL)

## TX Label: AvmTest/data_hashing_testing/0
- Total duration: `361.567 ms`
- Total mana used: `1,335,210`
- Mana per second: `3,692,838`
- Total instructions executed: `59,612`
- Private insertions:
    - Non-revertible: `5.148 ms`
    - Revertible: `0.672 ms`
- Proving:
    - Simulation (all): `375 ms`
    - Trace generation (all): `4,043 ms`
    - Trace generation interactions: `1,127 ms`
    - Trace generation traces: `2,896 ms`
- Enqueued public calls:
    - **Fn: AvmTest.assert_calldata_copy_large**
        - Duration: `155.984 ms`
        - Mana used: `612,309`
        - Mana per second: `3,925,472`
        - Instructions executed: `27,529`
    - **Fn: AvmTest.assert_calldata_copy**
        - Duration: `13.287 ms`
        - Mana used: `17,982`
        - Mana per second: `1,353,396`
        - Instructions executed: `808`
    - **Fn: AvmTest.assert_calldata_copy_large**
        - Duration: `97.561 ms`
        - Mana used: `612,309`
        - Mana per second: `6,276,144`
        - Instructions executed: `27,529`
    - **Fn: AvmTest.call_fee_juice**
        - Duration: `23.704 ms`
        - Mana used: `30,051`
        - Mana per second: `1,267,752`
        - Instructions executed: `1,217`
    - **Fn: AvmTest.call_auth_registry**
        - Duration: `20.407 ms`
        - Mana used: `30,762`
        - Mana per second: `1,507,427`
        - Instructions executed: `1,251`
    - **Fn: AvmTest.call_instance_registry**
        - Duration: `20.087 ms`
        - Mana used: `31,797`
        - Mana per second: `1,582,954`
        - Instructions executed: `1,278`
```

### Column Sizes
```
AvmCircuit (avm_prove) BB out - Column sizes per namespace: (mem: 2342.15 MiB)
  precomputed: 2097152 (~2^21) (mem: 2342.15 MiB)
  public: 4675 (~2^13) (mem: 2342.15 MiB)
  address_derivation: 4 (~2^2) (mem: 2342.15 MiB)
  alu: 5599 (~2^13) (mem: 2342.15 MiB)
  bc_decomposition: 75986 (~2^17) (mem: 2342.15 MiB)
  bc_hashing: 821 (~2^10) (mem: 2342.15 MiB)
  bc_retrieval: 10 (~2^4) (mem: 2342.15 MiB)
  bitwise: 0 (~2^0) (mem: 2342.15 MiB)
  calldata: 613 (~2^10) (mem: 2342.15 MiB)
[16:04:27.902] VERBOSE: avm-simulation-tester AvmCircuit (avm_prove) BB out - 
  class_id_derivation: 4 (~2^2) (mem: 2342.15 MiB)
  context: 3 (~2^2) (mem: 2342.15 MiB)
  contract_instance_retrieval: 10 (~2^4) (mem: 2342.15 MiB)
  data_copy: 1224 (~2^11) (mem: 2342.15 MiB)
  ecc: 1526 (~2^11) (mem: 2342.15 MiB)
  emit_unencrypted_log: 0 (~2^0) (mem: 2342.15 MiB)
  execution: 59613 (~2^16) (mem: 2342.15 MiB)
  ff_gt: 81 (~2^7) (mem: 2342.15 MiB)
  get_contract_instance: 0 (~2^0) (mem: 2342.15 MiB)
  gt: 61515 (~2^16) (mem: 2342.15 MiB)
  instr_fetching: 2258 (~2^12) (mem: 2342.15 MiB)
  internal_call: 1300 (~2^11) (mem: 2342.15 MiB)
  keccak_memory: 0 (~2^0) (mem: 2342.15 MiB)
  keccakf1600: 0 (~2^0) (mem: 2342.15 MiB)
  l1_to_l2_message_tree_check: 0 (~2^0) (mem: 2342.15 MiB)
  memory: 174131 (~2^18) (mem: 2342.15 MiB)
  merkle_check: 880 (~2^10) (mem: 2342.15 MiB)
  note_hash_tree_check: 0 (~2^0) (mem: 2342.15 MiB)
  nullifier_check: 7 (~2^3) (mem: 2342.15 MiB)
  poseidon2_hash: 2197 (~2^12) (mem: 2342.15 MiB)
  poseidon2_perm: 2199 (~2^12) (mem: 2342.15 MiB)
  protocol_contract: 3 (~2^2) (mem: 2342.15 MiB)
  public_data_check: 12 (~2^4) (mem: 2342.15 MiB)
  public_data_squash: 2 (~2^1) (mem: 2342.15 MiB)
  range_check: 63751 (~2^16) (mem: 2342.15 MiB)
  retrieved_bytecodes_tree_check: 18 (~2^5) (mem: 2342.15 MiB)
  scalar_mul: 1017 (~2^10) (mem: 2342.15 MiB)
  sha256: 0 (~2^0) (mem: 2342.15 MiB)
  to_radix: 1017 (~2^10) (mem: 2342.15 MiB)
  tx: 18 (~2^5) (mem: 2342.15 MiB)
  update_check: 6 (~2^3) (mem: 2342.15 MiB)
  written_public_data_slots_tree_check: 0 (~2^0) (mem: 2342.15 MiB)
  lookup: 75970 (~2^17) (mem: 2342.15 MiB)
  perm: 0 (~2^0) (mem: 2342.15 MiB)
Sum of all column rows: 30381356 (~2^25) (mem: 2342.15 MiB)
```
### `check_circuit` Timings
(`num threads: 16`)
```
Running check (with skippable) circuit over 196609 rows.
AvmCircuit (avm_check_circuit) BB out - proving/check_circuit_ms: 1449
proving/init_polys_to_be_shifted_ms: 31
proving/init_polys_unshifted_ms: 32
proving/prove:compute_polynomials_ms: 166
proving/set_polys_shifted_ms: 0
proving/set_polys_unshifted_ms: 102
simulation/all_ms: 371
tracegen/address_derivation_ms: 0
tracegen/all_ms: 4399
tracegen/alu_ms: 30
tracegen/bitwise_ms: 0
tracegen/bytecode_decomposition_ms: 2641
tracegen/bytecode_hashing_ms: 11
tracegen/bytecode_retrieval_ms: 0
tracegen/calldata_hashing_ms: 3
tracegen/calldata_retrieval_ms: 0
tracegen/class_id_derivation_ms: 0
tracegen/context_stack_ms: 0
tracegen/contract_instance_retrieval_ms: 0
tracegen/data_copy_ms: 23
tracegen/ecc_add_memory_ms: 0
tracegen/ecc_add_ms: 31
tracegen/emit_unencrypted_log_ms: 0
tracegen/execution_ms: 2595
tracegen/field_gt_ms: 0
tracegen/get_contract_instance_ms: 0
tracegen/gt_ms: 50
tracegen/instruction_fetching_ms: 18
tracegen/interactions_ms: 1686
tracegen/internal_call_stack_ms: 0
tracegen/keccak_f1600_memory_slices_ms: 0
tracegen/keccak_f1600_permutation_ms: 0
tracegen/l1_to_l2_message_tree_check_ms: 0
tracegen/memory_ms: 518
tracegen/merkle_check_ms: 21
tracegen/note_hash_tree_check_ms: 0
tracegen/nullifier_tree_check_ms: 0
tracegen/poseidon2_hash_ms: 17
tracegen/poseidon2_permutation_ms: 110
tracegen/poseidon2_permutation_with_memory_ms: 0
tracegen/protocol_contract_ms: 0
tracegen/public_data_tree_check_ms: 0
tracegen/range_check_ms: 176
tracegen/retrieved_bytecodes_tree_check_ms: 0
tracegen/scalar_mul_ms: 2
tracegen/sha256_compression_ms: 0
tracegen/to_radix_memory_ms: 0
tracegen/to_radix_ms: 37
tracegen/traces_ms: 2679
tracegen/tx_ms: 0
tracegen/update_check_ms: 0
tracegen/written_public_data_slots_tree_check_ms: 0
```

MirandaWood commented on noir-projects/noir-contracts/contracts/test/avm_test_contract/src/main.nr:
```diff
@@ -325,7 +328,7 @@ pub contract AvmTest {
     #[public]
     fn assert_calldata_copy(args: [Field; 3]) {
-        let offset = 0;
+        let offset = 1; // Skip the selector stored at i = 0
```
Note: I realised this call was always failing (regardless of my calldata changes), so this is a quick fix

fcarreiro reviewed (commented):
Interesting! Could you add the whole timing+trace sizes output to the PR description? (or reply to this comment) how many cpus are you using?

dbanks12 commented on yarn-project/simulator/src/public/fixtures/bulk_test.ts:
```diff
@@ -144,3 +144,74 @@ export async function megaBulkTest(
+
+// Runs multiple enqueued calls which perform calldata hashing, calldata copy, and bytecode hashing:
+export async function dataHashingBulk(
+  tester: PublicTxSimulationTester,
```
Another test that I think would be good (can probably be added later during worst-case testing) would be one that generates ~21 bytecodes of max length and calls them all (via enqueued or nested calls). That number 21 is our const `MAX_PUBLIC_CALLS_TO_UNIQUE_CONTRACT_CLASS_IDS`. So 21 max-length bytecodes should fill up our bytecode traces to the max.

dbanks12 commented on yarn-project/bb-prover/src/avm_proving_tests/avm_data_hashing_bulk.test.ts:
```diff
@@ -0,0 +1,42 @@
+
+const TIMEOUT = 180_000;
+
+describe('AVM data hashing bulk test', () => {
+  const logger = createLogger('avm-data-hashing-bulk-test');
```
This test feels "bulky" so i think naming is fine, but just for future reference, note that there are many other proving tests in this directory that we don't name "bulk". Today, the only ones named "bulk" call the `bulk_testing` function on the Noir test contract.

MirandaWood replied:
Good point - I will actually rename this because it bothers me now 😅

dbanks12 commented on yarn-project/bb-prover/src/avm_proving_tests/avm_data_hashing.test.ts:
```diff
@@ -0,0 +1,42 @@
+    // check_circuit only for now (~45s proving test time)
+    tester = await AvmProvingTester.new(/*checkCircuitOnly=*/ true, /*globals=*/ defaultGlobals(), metrics);
+  });
+
```
Oof.... it takes 45s to prove in checkCircuit mode.... That's brutal.

MirandaWood replied:
Ah this meant /not/ in checkCircuit mode (45s full proving) - now it's 25 since I removed the spammy nested calls, and around 14s in checkCircuit mode

dbanks12 commented on noir-projects/noir-contracts/contracts/test/avm_test_contract/src/main.nr:
```diff
@@ -325,8 +328,8 @@ pub contract AvmTest {
     #[public]
-    fn assert_calldata_copy(args: [Field; 3]) {
-        let offset = 0;
+    fn assert_calldata_copy(args: [Field; 3], with_selector: bool) {
+        let offset = with_selector as u32; // Skip the selector stored at i = 0 if used
         let cd: [Field; 3] = dep::aztec::context::public_context::calldata_copy(offset, 3);
         assert(cd == args, "Calldata copy failed");
     }
```
I'm confused about `with_selector`. If it is false/0, then we start copying at calldata offset=0, meaning we'll copy the selector. If it is true/1, then we start copying at calldata offset=1, meaning we'll skip the selector, right?

Maybe just remove the `with_selector` arg and use offset 1 with a comment ("offset of 1 to skip function selector").

dbanks12 replied:
But maybe I'm just confused about this arg and what it means....

MirandaWood replied:
Ah I think my naming here is not great, sorry! If the calldata includes the prepended selector, `with_selector` = 1. Otherwise it's 0 (this happens in some simulation testing where we don't go via dispatch), and we don't offset. 

I originally changed it to 1, but this actually broke the simulation tests, so I needed to have something that set offset as 1 for most tests (using dispatch), and as 0 for simulating.

dbanks12 commented on yarn-project/simulator/src/public/avm/fixtures/base_avm_simulation_tester.ts:
```diff
@@ -88,6 +90,36 @@ export abstract class BaseAvmSimulationTester {
     return feeJuice.instance;
   }
 
+  async registerAuthContract(): Promise<ContractInstanceWithAddress> {
+    const authRegistry = await getCanonicalAuthRegistry();
+    const authRegistryContractClassPublic = {
+      ...authRegistry.contractClass,
+      privateFunctions: [],
+      utilityFunctions: [],
+    };
+    await this.contractDataSource.addNewContract(
+      authRegistry.artifact,
+      authRegistryContractClassPublic,
+      authRegistry.instance,
+    );
+    return authRegistry.instance;
+  }
+
+  async registerInstanceRegistryContract(): Promise<ContractInstanceWithAddress> {
+    const instanceRegistry = await getCanonicalInstanceRegistry();
+    const instanceRegistryContractClassPublic = {
+      ...instanceRegistry.contractClass,
+      privateFunctions: [],
+      utilityFunctions: [],
+    };
+    await this.contractDataSource.addNewContract(
+      instanceRegistry.artifact,
+      instanceRegistryContractClassPublic,
+      instanceRegistry.instance,
+    );
+    return instanceRegistry.instance;
+  }
```
Cool! Thanks for adding this.

fcarreiro commented on yarn-project/simulator/src/public/fixtures/bulk_test.ts:
```diff
@@ -144,3 +144,74 @@ export async function megaBulkTest(
+        address: avmTestContract.address,
+        fnName: 'nested_call_large_calldata',
+        args: [Array.from({ length: 300 }, () => Fr.random())],
+      },
```
I think these are probably the "bulk" of the test. Do we really need them and/or do we really need them to be as big? 90% of the work on each of these is (IIUC) exercising the calldatacopy mechanisms for nested calls, which is not what you've just implemented (really it's in Noir, and we probably have/should have some test calling this?)

Would it be possible to remove these 3 and instead create a `assert_calldata_copy_large` which, well, does the same as the normal one but with 300 fields? Then you can do
* assert_calldata_copy_large
* assert_calldata_copy
* assert_calldata_copy_large

I'm pretty sure if we did that, keeping the 3 external calls to the contracts, then maybe it would be small enough to (fingers crossed) add to the normal bulk test.

In any case maybe worth experimenting?

MirandaWood replied:
I'd be happy to remove them! I just wanted something that did calldatacopy and wasn't just a couple of values, since this looks up the calldata size which I constrained with the hashing work. I kind of wanted to 'spam' and put this through its paces but maybe this is not the place for that kind of test.

IIUC you're suggesting I remove the 3x `nested_call_large_calldata` calls, and replace two of the three `assert_calldata_copy` calls with a new `assert_calldata_copy_large` call? Then move everything to the normal bulk test, or keep two tests (one normal bulk with one call, now including `assert_calldata_copy_large` nested call, plus this hashing 'bulk')?

MirandaWood replied:
I removed the nested calls and updated the existing `assert_calldata` calls to include the new larger tests - if this is what you were suggesting I'll update the stats on the PR! With `check_circuit_only = true` it's now around 14s, and 26s for proving. Otherwise lmk what I've misunderstood and happy to implement :)

MirandaWood committed:
feat(avm): Bulk test for bytecode/calldata hashing

## Hashing Bulk Test
Originally part of #17224, but I ~separated it out because the test is pretty long (and to help with reviewing!).~

UPDATE: No longer a separate test but included in separate enqueued calls in the main bulk test. This 'only' increases the time by about 10s. The below stats refer to the data hashing tests on their own!

I added ~a new bulk-style test with~ multiple enqueued calls performing calldata and bytecode hashing.

Key points:

- Performs 3x calldata hashes with reads
- Performs 3x protocol contract calls (=> 3x bytecode hashes)
- Currently takes ~20s (10s check_circuit only)

Since we dedupe hashing events, I added different protocol contracts to force the bytecodes to be unique.

### Metrics

```
INFO: avm-data-hashing-test # Public TX Simulation Metrics (ALL)

## TX Label: AvmTest/data_hashing_testing/0
- Total duration: `361.567 ms`
- Total mana used: `1,335,210`
- Mana per second: `3,692,838`
- Total instructions executed: `59,612`
- Private insertions:
    - Non-revertible: `5.148 ms`
    - Revertible: `0.672 ms`
- Proving:
    - Simulation (all): `375 ms`
    - Trace generation (all): `4,043 ms`
    - Trace generation interactions: `1,127 ms`
    - Trace generation traces: `2,896 ms`
- Enqueued public calls:
    - **Fn: AvmTest.assert_calldata_copy_large**
        - Duration: `155.984 ms`
        - Mana used: `612,309`
        - Mana per second: `3,925,472`
        - Instructions executed: `27,529`
    - **Fn: AvmTest.assert_calldata_copy**
        - Duration: `13.287 ms`
        - Mana used: `17,982`
        - Mana per second: `1,353,396`
        - Instructions executed: `808`
    - **Fn: AvmTest.assert_calldata_copy_large**
        - Duration: `97.561 ms`
        - Mana used: `612,309`
        - Mana per second: `6,276,144`
        - Instructions executed: `27,529`
    - **Fn: AvmTest.call_fee_juice**
        - Duration: `23.704 ms`
        - Mana used: `30,051`
        - Mana per second: `1,267,752`
        - Instructions executed: `1,217`
    - **Fn: AvmTest.call_auth_registry**
        - Duration: `20.407 ms`
        - Mana used: `30,762`
        - Mana per second: `1,507,427`
        - Instructions executed: `1,251`
    - **Fn: AvmTest.call_instance_registry**
        - Duration: `20.087 ms`
        - Mana used: `31,797`
        - Mana per second: `1,582,954`
        - Instructions executed: `1,278`
```

### Column Sizes
```
AvmCircuit (avm_prove) BB out - Column sizes per namespace: (mem: 2342.15 MiB)
  precomputed: 2097152 (~2^21) (mem: 2342.15 MiB)
  public: 4675 (~2^13) (mem: 2342.15 MiB)
  address_derivation: 4 (~2^2) (mem: 2342.15 MiB)
  alu: 5599 (~2^13) (mem: 2342.15 MiB)
  bc_decomposition: 75986 (~2^17) (mem: 2342.15 MiB)
  bc_hashing: 821 (~2^10) (mem: 2342.15 MiB)
  bc_retrieval: 10 (~2^4) (mem: 2342.15 MiB)
  bitwise: 0 (~2^0) (mem: 2342.15 MiB)
  calldata: 613 (~2^10) (mem: 2342.15 MiB)
[16:04:27.902] VERBOSE: avm-simulation-tester AvmCircuit (avm_prove) BB out -
  class_id_derivation: 4 (~2^2) (mem: 2342.15 MiB)
  context: 3 (~2^2) (mem: 2342.15 MiB)
  contract_instance_retrieval: 10 (~2^4) (mem: 2342.15 MiB)
  data_copy: 1224 (~2^11) (mem: 2342.15 MiB)
  ecc: 1526 (~2^11) (mem: 2342.15 MiB)
  emit_unencrypted_log: 0 (~2^0) (mem: 2342.15 MiB)
  execution: 59613 (~2^16) (mem: 2342.15 MiB)
  ff_gt: 81 (~2^7) (mem: 2342.15 MiB)
  get_contract_instance: 0 (~2^0) (mem: 2342.15 MiB)
  gt: 61515 (~2^16) (mem: 2342.15 MiB)
  instr_fetching: 2258 (~2^12) (mem: 2342.15 MiB)
  internal_call: 1300 (~2^11) (mem: 2342.15 MiB)
  keccak_memory: 0 (~2^0) (mem: 2342.15 MiB)
  keccakf1600: 0 (~2^0) (mem: 2342.15 MiB)
  l1_to_l2_message_tree_check: 0 (~2^0) (mem: 2342.15 MiB)
  memory: 174131 (~2^18) (mem: 2342.15 MiB)
  merkle_check: 880 (~2^10) (mem: 2342.15 MiB)
  note_hash_tree_check: 0 (~2^0) (mem: 2342.15 MiB)
  nullifier_check: 7 (~2^3) (mem: 2342.15 MiB)
  poseidon2_hash: 2197 (~2^12) (mem: 2342.15 MiB)
  poseidon2_perm: 2199 (~2^12) (mem: 2342.15 MiB)
  protocol_contract: 3 (~2^2) (mem: 2342.15 MiB)
  public_data_check: 12 (~2^4) (mem: 2342.15 MiB)
  public_data_squash: 2 (~2^1) (mem: 2342.15 MiB)
  range_check: 63751 (~2^16) (mem: 2342.15 MiB)
  retrieved_bytecodes_tree_check: 18 (~2^5) (mem: 2342.15 MiB)
  scalar_mul: 1017 (~2^10) (mem: 2342.15 MiB)
  sha256: 0 (~2^0) (mem: 2342.15 MiB)
  to_radix: 1017 (~2^10) (mem: 2342.15 MiB)
  tx: 18 (~2^5) (mem: 2342.15 MiB)
  update_check: 6 (~2^3) (mem: 2342.15 MiB)
  written_public_data_slots_tree_check: 0 (~2^0) (mem: 2342.15 MiB)
  lookup: 75970 (~2^17) (mem: 2342.15 MiB)
  perm: 0 (~2^0) (mem: 2342.15 MiB)
Sum of all column rows: 30381356 (~2^25) (mem: 2342.15 MiB)
```
### `check_circuit` Timings
(`num threads: 16`)
```
Running check (with skippable) circuit over 196609 rows.
AvmCircuit (avm_check_circuit) BB out - proving/check_circuit_ms: 1449
proving/init_polys_to_be_shifted_ms: 31
proving/init_polys_unshifted_ms: 32
proving/prove:compute_polynomials_ms: 166
proving/set_polys_shifted_ms: 0
proving/set_polys_unshifted_ms: 102
simulation/all_ms: 371
tracegen/address_derivation_ms: 0
tracegen/all_ms: 4399
tracegen/alu_ms: 30
tracegen/bitwise_ms: 0
tracegen/bytecode_decomposition_ms: 2641
tracegen/bytecode_hashing_ms: 11
tracegen/bytecode_retrieval_ms: 0
tracegen/calldata_hashing_ms: 3
tracegen/calldata_retrieval_ms: 0
tracegen/class_id_derivation_ms: 0
tracegen/context_stack_ms: 0
tracegen/contract_instance_retrieval_ms: 0
tracegen/data_copy_ms: 23
tracegen/ecc_add_memory_ms: 0
tracegen/ecc_add_ms: 31
tracegen/emit_unencrypted_log_ms: 0
tracegen/execution_ms: 2595
tracegen/field_gt_ms: 0
tracegen/get_contract_instance_ms: 0
tracegen/gt_ms: 50
tracegen/instruction_fetching_ms: 18
tracegen/interactions_ms: 1686
tracegen/internal_call_stack_ms: 0
tracegen/keccak_f1600_memory_slices_ms: 0
tracegen/keccak_f1600_permutation_ms: 0
tracegen/l1_to_l2_message_tree_check_ms: 0
tracegen/memory_ms: 518
tracegen/merkle_check_ms: 21
tracegen/note_hash_tree_check_ms: 0
tracegen/nullifier_tree_check_ms: 0
tracegen/poseidon2_hash_ms: 17
tracegen/poseidon2_permutation_ms: 110
tracegen/poseidon2_permutation_with_memory_ms: 0
tracegen/protocol_contract_ms: 0
tracegen/public_data_tree_check_ms: 0
tracegen/range_check_ms: 176
tracegen/retrieved_bytecodes_tree_check_ms: 0
tracegen/scalar_mul_ms: 2
tracegen/sha256_compression_ms: 0
tracegen/to_radix_memory_ms: 0
tracegen/to_radix_ms: 37
tracegen/traces_ms: 2679
tracegen/tx_ms: 0
tracegen/update_check_ms: 0
tracegen/written_public_data_slots_tree_check_ms: 0
```
